mlir/lib/Dialect/Tosa/Transforms/SinkInputOpsThroughConcat.cpp

```diff
@@ -0,0 +1,430 @@
+            concat, "Check for broadcast on an unshaped or not static type.");
+      }
+    }
+    assert(shapes.size() == op->getNumOperands() &&
+           "Something went wrong with the above loop.");
+
+    // check that the ranks on the axis that the concat uses is not one for all
+    // ops or one for all ops
+    const auto axis = concat.getAxis();
+    const size_t oneDimensions =
+        llvm::count_if(shapes, [&](const auto &s) { return s[axis] == 1; });
+    if (oneDimensions != shapes.size() && oneDimensions != 0) {
+      return rewriter.notifyMatchFailure(
+          concat, "Operand broadcasts on same axis then concat.");
+    }
+    return success();
+  }
+};
+
+template <class OpT>
+struct SinkReduceOp : public SinkSpecificOp<OpT> {
+  using SinkSpecificOp<OpT>::SinkSpecificOp;
+
+protected:
+  // check that the concat happens on another axis than this one
+  LogicalResult doExtraChecks(OpT op, tosa::ConcatOp concat,
+                              PatternRewriter &rewriter) const override {
+    if (op.getAxis() == concat.getAxis())
+      return rewriter.notifyMatchFailure(
+          concat, "Operator must not be on the same axis than concat.");
+    return success();
+  }
+};
+
+struct SinkMatmulOp : public SinkSpecificOp<tosa::MatMulOp> {
+  using SinkSpecificOp<tosa::MatMulOp>::SinkSpecificOp;
+
+protected:
+  // check that the concat happens on another axis than the one used by matmul
+  LogicalResult doExtraChecks(tosa::MatMulOp op, tosa::ConcatOp concat,
+                              PatternRewriter &rewriter) const override {
+    if (concat.getAxis() != 0)
+      return rewriter.notifyMatchFailure(
+          concat, "Matmul concat about different axis not yet supported");
+    return success();
+  }
+};
+
+struct SinkReshapeOp : public SinkSpecificOp<tosa::ReshapeOp> {
+  using SinkSpecificOp<tosa::ReshapeOp>::SinkSpecificOp;
+
+  LogicalResult matchAndRewrite(tosa::ConcatOp concatOp,
+                                PatternRewriter &rewriter) const override {
+    auto reshapeOrError = doGenericChecks(concatOp, rewriter);
+    if (failed(reshapeOrError))
+      return reshapeOrError;
+    auto reshape = *reshapeOrError;
+
+    assert(reshape->getNumOperands() == 1 &&
+           "shape must have exactly one operand");
+    const auto opType = reshape->getOperandTypes().front();
+    const auto tenType = dyn_cast<ShapedType>(opType);
+    if (!tenType || !tenType.hasStaticShape())
+      return rewriter.notifyMatchFailure(
+          concatOp, "Dynamic shapes for reshapes are not supported.");
+    const auto oldShape = tenType.getShape();
+    const auto newShape = reshape.getNewShape();
+
+    const auto newConcatAxis = concatOp.getAxis();
+    int64_t newPrefixProduct = 1;
+    for (size_t i = 0; i < newConcatAxis; ++i) {
+      newPrefixProduct *= newShape[i];
+    }
+
+    int64_t oldPrefixProduct = 1;
+    size_t oldConcatAxis = oldShape.size(); // default to invalid value
```

Comment:
I find choosing `oldShape.size()` as "invalid state" confusing. What do you thing of:
```suggestion
    std::optional<int64_t> oldConcatAxis;
```

Reply:
I'm fine with it. Will change.